cornac/eval_methods/next_basket_evaluation.py

```diff
@@ -0,0 +1,436 @@
+        If `True`, unknown items will be ignored during model evaluation.
+
+    verbose: bool, optional, default: False
+        Output running log.
+
+    """
+
+    def __init__(
+        self,
+        data=None,
+        test_size=0.2,
+        val_size=0.0,
+        fmt="UBI",
+        seed=None,
+        repetition_eval=False,
+        exploration_eval=False,
+        exclude_unknowns=True,
+        verbose=False,
+        **kwargs
+    ):
+        super().__init__(
+            data=data,
+            fmt=fmt,
+            seed=seed,
+            exclude_unknowns=exclude_unknowns,
+            verbose=verbose,
+            **kwargs
+        )
+        self.repetition_eval = repetition_eval
+        self.exploration_eval = exploration_eval
+
+        self.train_size, self.val_size, self.test_size = self.validate_size(
+            val_size, test_size, len(set(u for (u, *_) in self._data))
+        )
+        self._split()
+
+    @staticmethod
+    def validate_size(val_size, test_size, num_users):
+        if val_size is None:
+            val_size = 0.0
+        elif val_size < 0:
+            raise ValueError("val_size={} should be greater than zero".format(val_size))
+        elif val_size >= num_users:
+            raise ValueError(
+                "val_size={} should be less than the number of users {}".format(
+                    val_size, num_users
+                )
+            )
+
+        if test_size is None:
+            test_size = 0.0
+        elif test_size < 0:
+            raise ValueError(
+                "test_size={} should be greater than zero".format(test_size)
+            )
+        elif test_size >= num_users:
+            raise ValueError(
+                "test_size={} should be less than the number of users {}".format(
+                    test_size, num_users
+                )
+            )
+
+        if val_size < 1:
+            val_size = ceil(val_size * num_users)
+        if test_size < 1:
+            test_size = ceil(test_size * num_users)
+
+        if val_size + test_size >= num_users:
+            raise ValueError(
+                "The sum of val_size and test_size ({}) should be smaller than the number of users {}".format(
+                    val_size + test_size, num_users
+                )
+            )
+
+        train_size = num_users - (val_size + test_size)
+
+        return int(train_size), int(val_size), int(test_size)
+
+    def _split(self):
+        users = list(set(u for(u, *_) in self._data))
+        data_idx = self.rng.permutation(len(users))
+        train_idx = data_idx[: self.train_size]
+        test_idx = data_idx[-self.test_size :]
+        val_idx = data_idx[self.train_size : -self.test_size]
+
+        train_users = safe_indexing(users, train_idx)
+        test_users = safe_indexing(users, test_idx)
+        val_users = safe_indexing(users, val_idx) if len(val_idx) > 0 else None
+
+        train_data = [tup for tup in self._data if tup[0] in train_users]
```

Comment:
@tqtg We need a separate `_split` method to split data based on users. However, this line of code is not efficient.